### ISSUE TITLE
<base href> breaks fragment links

### ISSUE THREAD

**Issue by delan** [delan/autost#17]
the threads page template uses &lt;base href>, but this breaks fragment links in posts.

for example, say you’re in /posts/1.html and you have:

```html
<base href="/posts/">
<a href="#user-content-usb3sun-user-guide">
```

that becomes this, which may work but will navigate for no reason:

```html
<a href="/posts/#user-content-usb3sun-user-guide">
```

**Comment by delan** [comment]
- <https://stackoverflow.com/questions/1889076/what-are-the-recommendations-for-html-base-tag>
- <https://www.drupal.org/node/13148>
- <https://mademistakes.com/mastering-jekyll/site-url-baseurl/>
- <https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/path-prefix/>
- <https://www.11ty.dev/docs/plugins/html-base/>
- <https://www.11ty.dev/docs/filters/url/>

option 1: use js to fix fragment links from `#foo` to `url#foo`
- if done correctly, does not require navigation when clicking links
- but requires js :(

option 2: opt-in fix with a `{{ |url }}` filter
- easiest to understand, and the most common solution
- delegates decision about whether to fix to the author
- need to be careful to avoid triggering it in (newly) imported posts
- any syntax we use may break existing posts, needing a migration

option 3: magic fix
- requires parsing all attributes that can take urls (we already have this kinda)
- strictly speaking, requires parsing all css `url()` values (we don’t have this)
- requires us to decide what kinds of urls to fix, which seems tricky
  - `http://host/path?query#fragment` → don’t fix
  - `//host/path?query#fragment` → don’t fix
  - `/path?query#fragment` → do we fix this or not?
  - `path?query#fragment` → do we fix this or not?
  - `?query#fragment` → how about this? (less important)
  - `#fragment` → don’t fix

**Comment by delan** [comment]
if we go with a magic fix, what kinds of urls do we fix? say we have `base_url = "/posts/"`…

| if we fix this →<br>then we link to this ↓ with | `/path/to/file` only                                | `path/to/file` only | both                                                |
| ----------------------------------------------- | --------------------------------------------------- | ------------------- | --------------------------------------------------- |
| `/posts/2.html`                                 | `/2.html`                                           | `2.html`            | either                                              |
| `/cats/diffie`                                  | `[../]../cats/diffie` :(<br>`//host/cats/diffie` :( | `/cats/diffie`      | `[../]../cats/diffie` :(<br>`//host/cats/diffie` :( |
| `/posts/tagged/cats.html`                       | `/tagged/cats.html`                                 | `tagged/cats.html`  | either                                              |

if we fix `/path/to/file` urls only…
- makes more sense, according to @ar1a [and 11ty](https://www.11ty.dev/docs/plugins/html-base/)
- can’t reliably link out of base dir without also locking in a host
- breaks existing posts, so we would need to migrate and/or reimport

| current url →<br>link url ↓ | `/posts/1.html`         | `/posts/tagged/cats.html` |
| --------------------------- | ----------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/    | `/posts/tagged/2.html` :( |
| `/2.html`                   | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/cats/diffie`              | `/posts/cats/diffie` :( | `/posts/cats/diffie` :(   |

if we fix `path/to/file` urls only…
- can break out of base dir with leading `/`
- fixes fragment links without breaking any other links

| current url →<br>link url ↓ | `/posts/1.html`      | `/posts/tagged/cats.html` |
| --------------------------- | -------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/ | `/posts/2.html` \\o/      |
| `/2.html`                   | `/2.html` :(         | `/2.html` :(              |
| `/cats/diffie`              | `/cats/diffie` \\o/  | `/cats/diffie` \\o/       |

if we fix both…
- can’t reliably link out of base dir without also locking in a host
- breaks existing posts that link out of base dir

| current url →<br>link url ↓ | `/posts/1.html`         | `/posts/tagged/cats.html` |
| --------------------------- | ----------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/2.html`                   | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/cats/diffie`              | `/posts/cats/diffie` :( | `/posts/cats/diffie` :(   |

**Comment by delan** [comment]
if we go with a magic fix, what kinds of urls do we fix? say we have `base_url = "/posts/"`…

| if we fix this →<br>then we link to this ↓ with | `/path/to/file` only                                | `path/to/file` only | both                                                |
| ----------------------------------------------- | --------------------------------------------------- | ------------------- | --------------------------------------------------- |
| `/posts/2.html`                                 | `/2.html`                                           | `2.html`            | either                                              |
| `/cats/diffie`                                  | `[../]../cats/diffie` :(<br>`//host/cats/diffie` :( | `/cats/diffie`      | `[../]../cats/diffie` :(<br>`//host/cats/diffie` :( |
| `/posts/tagged/cats.html`                       | `/tagged/cats.html`                                 | `tagged/cats.html`  | either                                              |

if we fix `/path/to/file` urls only…
- makes more sense, according to @ar1a [and 11ty](https://www.11ty.dev/docs/plugins/html-base/)
- can’t reliably link out of base dir without also locking in a host
- breaks existing posts with `attachments/...` urls
- breaks existing posts that link out of base dir
- so we would need to migrate and/or reimport posts

| current url →<br>link url ↓ | `/posts/1.html`         | `/posts/tagged/cats.html` |
| --------------------------- | ----------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/    | `/posts/tagged/2.html` :( |
| `/2.html`                   | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/cats/diffie`              | `/posts/cats/diffie` :( | `/posts/cats/diffie` :(   |

if we fix `path/to/file` urls only…
- can link out of base dir with leading `/`
- fixes fragment links without breaking any other links

| current url →<br>link url ↓ | `/posts/1.html`      | `/posts/tagged/cats.html` |
| --------------------------- | -------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/ | `/posts/2.html` \\o/      |
| `/2.html`                   | `/2.html` :(         | `/2.html` :(              |
| `/cats/diffie`              | `/cats/diffie` \\o/  | `/cats/diffie` \\o/       |

if we fix both…
- can’t reliably link out of base dir without also locking in a host
- breaks existing posts that link out of base dir
- so we would need to migrate and/or reimport posts

| current url →<br>link url ↓ | `/posts/1.html`         | `/posts/tagged/cats.html` |
| --------------------------- | ----------------------- | ------------------------- |
| `2.html`                    | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/2.html`                   | `/posts/2.html` \\o/    | `/posts/2.html` \\o/      |
| `/cats/diffie`              | `/posts/cats/diffie` :( | `/posts/cats/diffie` :(   |

**Comment by delan** [comment]
you know your change is Good when it even “fixes” a couple of fucked up chosts :)

![image](https://github.com/user-attachments/assets/8af29f3e-6050-42e2-8603-e2200a35a44c)

**Comment by delan** [comment]
fixed in d14767d0fe1ab13e01ec827cecdc4348d7331fda through f5415d8312a7e096109448e33010f6ef5bb30d31; 4e484df4221efb980741d2299bbc4ce31bb9d936 unfortunately comes with a pretty big perf regression, but i’ve managed to reclaim most of the runtime in subsequent commits:
- 1.107s feed template: add <author> and <category> for each thread (3c55b7200f6f5106a96c38704e201508f50bb34f)
- 1.131s remove reliance on <base href> in several other places (f00b1199217bb8e7e6b93cc261eb2e75994d97b8)
- 4.070s guarantee that our html output fixes relative urls (4e484df4221efb980741d2299bbc4ce31bb9d936)
- 4.057s remove stray comment in TemplatedPost::filter (f5415d8312a7e096109448e33010f6ef5bb30d31)
- 4.130s [nfc] refactor fix_relative_urls() (c07837bd9944b62755ae84edf596b02eadd0ef9d)
- 2.968s perf: don’t fix relative urls twice in threads pages (13b4e0e8089070e54a7aef7e472477afb5a36b26)
- 2.951s perf: scan only specific attributes for attachment urls (fddfb1f347b49b040981a1c82d2553cada91aa07)
- 3.113s perf: prepare to parallelise render process (f82f5dd0d048a54f9132ee73c4c223e4fe975276)
- 3.161s perf: cache ThreadsContentTemplate from individual threads (05b939fc96fdc8f7c9412b73a44cbf3225a149d6)
- 2.869s perf: don’t fix relative urls twice in threads pages (again) (68ec5783f94aa2bae41d640ba9b127710a06d986)
- 2.846s perf: reduce clones in render process (part 1) (99b9d2aebd1dfed0d789671de81718ef3f35ae26)
